bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitor.java

```diff
@@ -0,0 +1,185 @@
+ * necessary to ensure data integrity: if updates to claims are processed out of
+ * order, the data will end up in the wrong final state.
+ * </p>
+ */
+public final class DataSetMonitor {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DataSetMonitor.class);
+
+	private final String bucketName;
+	private final int scanRepeatDelay;
+	private final DataSetProcessor dataSetProcessor;
+
+	private ScheduledExecutorService dataSetWatcherService;
+	private ScheduledFuture<?> dataSetWatcherFuture;
+
+	/**
+	 * Constructs a new {@link DataSetMonitor} instance. Note that this must be used
+	 * as a singleton service in the application: only one instance running at a
+	 * time is supported.
+	 * 
+	 * @param bucketName
+	 *            the name of the AWS S3 bucket to monitor
+	 * @param scanRepeatDelay
+	 *            the number of milliseconds to wait after completing one
+	 *            poll/process operation and starting another
+	 * @param dataSetProcessor
+	 *            the {@link DataSetProcessor} responsible for processing each
+	 *            of the {@link RifFilesEvent}s that are generated by this
+	 *            {@link DataSetMonitor}
+	 */
+	public DataSetMonitor(String bucketName, int scanRepeatDelay, DataSetProcessor dataSetProcessor) {
+		this.bucketName = bucketName;
+		this.scanRepeatDelay = scanRepeatDelay;
+		this.dataSetProcessor = dataSetProcessor;
+	}
+
+	/**
+	 * Starts this monitor: it will begin regularly polling for and processing
+	 * new data sets on a background thread. This particular method will return
+	 * immediately, but that background processing will continue to run
+	 * asynchronously until {@link #stop()} is called.
+	 */
+	public void start() {
+		this.dataSetWatcherService = Executors.newSingleThreadScheduledExecutor();
+		Runnable dataSetWatcher = new DataSetMonitorWorker(bucketName, dataSetProcessor);
+
+		/*
+		 * This kicks off the data set watcher, which will be run periodically
+		 * (as configured), until this instance's stop() method is called to
+		 * request a graceful shutdown. In turn, it basically acts as the
+		 * application's main loop, and will drive all of the application's data
+		 * processing, etc. on the ScheduledExecutorService's separate thread.
+		 * (To be clear: this method does not block, and should return normally
+		 * almost immediately.)
+		 */
+		this.dataSetWatcherFuture = dataSetWatcherService.scheduleWithFixedDelay(dataSetWatcher, 0, scanRepeatDelay,
+				TimeUnit.MILLISECONDS);
+	}
+
+	/**
+	 * <p>
+	 * Stops this monitor: it will wait for the current polling/processing
+	 * operation (if any) to complete normally, then stop any future
+	 * polling/processing runs from starting. This method will block until any
+	 * current operations have completed and future ones have been cancelled.
+	 * </p>
+	 * <p>
+	 * <strong>Note:</strong> This might block for a while! Some data sets have
+	 * terrabytes of data, and there is no way to safely stop processing in the
```

Comment:
Do you anticipate we will be able to pick up where we left off in the middle of processing a data set, for instance in the case of some kind of fatal error in the middle of processing?